src/Models/Cart.php

```diff
@@ -82,6 +86,48 @@ public function broadcastWith(): array
         return ['id' => $this->id];
     }
 
+    public function addItems(array|int $products): static
+    {
+        $products = Arr::wrap(is_array($products) && ! array_is_list($products) ? [$products] : $products);
+
+        foreach ($products as $product) {
+            $productModel = null;
+            if ($productId = is_array($product) ? data_get($product, 'id') : $product) {
+                $productModel = resolve_static(Product::class, 'query')
+                    ->whereKey($productId)
+                    ->first();
+            }
+
+            $data = [
+                'cart_id' => $this->id,
+                'product_id' => data_get($product, 'id', $productModel?->id),
+                'name' => data_get($product, 'name', $productModel?->name),
+                'amount' => $product['amount'] ?? 1,
+                'price' => $product['price']
+                    ?? PriceHelper::make($productModel)
+                        ->when(
+                            auth()->user() instanceof Address,
+                            fn (PriceHelper $price) => $price->setContact(auth()->user()->contact)
+                        )
+                        ->price()
+                        ?->price,
+            ];
+
+            if ($cartItem = $this->cartItems()->where('product_id', $productId)->first()) {
+                $data['id'] = $cartItem->id;
+                $data['amount'] = bcadd($cartItem->amount, $data['amount']);
+
+                $action = UpdateCartItem::make($data);
+            } else {
+                $action = CreateCartItem::make($data);
+            }
+
+            $action->validate()->checkPermission()->execute();
```

Comment:
first checkPermission() then validate(). There's no need to validate the data if the user does not have the right to do so.